### PR TITLE
Correctly handle newlines after/before comments

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This issue fixes the removal of empty lines between a leading comment and the previous statement:

```python
a  = 20

# leading comment
b = 10
```

Ruff removed the empty line between `a` and `b` because:
* The leading comments formatting does not preserve leading newlines (to avoid adding new lines at the top of a body)
* The `JoinNodesBuilder` counted the lines before `b`, which is 1 -> Doesn't insert a new line

This is fixed by changing the `JoinNodesBuilder` to count the lines instead *after* the last node. This correctly gives 1, and the `# leading comment` will insert the empty lines between any other leading comment or the node.

## Test Plan

I added a new test for empty lines.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4862** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4862" target="_blank">Graphite</a> 
    * **PR #4895** <a href="https://app.graphite.dev/github/pr/charliermarsh/ruff/4895" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/4895?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/trivia.rs [hunk 1]
```diff
@@ -146,6 +146,33 @@ pub(crate) fn lines_after(offset: TextSize, code: &str) -> u32 {
...
+    let mut iter = rest.char_indices();
+
+    while let Some((relative_offset, c)) = iter.next() {
+        match c {
+            '\n' | '\r' => return offset + TextSize::try_from(relative_offset).unwrap(),
+            '#' => {
+                // Skip the comment
+                let newline_offset = iter
+                    .as_str()
+                    .find(['\n', '\r'])
```
Comment (konstin) [inline crates/ruff_python_formatter/src/trivia.rs:161]: Does this handle CR LF correctly?

**Review by konstin — APPROVED** [review]

**Comment by MichaReiser** [comment]
@MichaReiser started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/4895).

**Comment by MichaReiser** [comment]
@MichaReiser merged this pull request with [Graphite](https://app.graphite.dev/github/pr/charliermarsh/ruff/4895).

**Pull request merged**